### ISSUE TITLE
Server has gone away - connecting with mariaDB

### ISSUE THREAD

**Issue by GiaBen** [zopefoundation/Products.ZMySQLDA#34]
## BUG/PROBLEM REPORT / FEATURE REQUEST

File "/home/myusername/apps/zope5plus/zope5100/eggs/transaction-4.0-py3.11.egg/transaction/_transaction.py", line 570, in abort
    raise v.with_traceback(tb)
  File "/home/myusername/apps/zope5plus/zope5100/eggs/transaction-4.0-py3.11.egg/transaction/_transaction.py", line 551, in abort
    rm.abort(self)
  File "/home/myusername/apps/zope5plus/zope5100/eggs/Products.ZSQLMethods-4.1-py3.11.egg/Shared/DC/ZRDB/TM.py", line 90, in abort
    self._abort()
  File "/home/myusername/apps/zope5plus/zope5100/eggs/Products.ZMySQLDA-5.0-py3.11.egg/Products/ZMySQLDA/db.py", line 679, in _abort
    self._query('ROLLBACK')
  File "/home/myusername/apps/zope5plus/zope5100/eggs/Products.ZMySQLDA-5.0-py3.11.egg/Products/ZMySQLDA/db.py", line 519, in _query
    self.db.query(query)
  File "/home/myusername/apps/zope5plus/zope5100/lib/python3.11/site-packages/MySQLdb/connections.py", line 261, in query
    _mysql.connection.query(self, query)
MySQLdb.OperationalError: (2006, 'Server has gone away')

### What I did:

I was just querying the mariaDB with a basic script like select * from table where id = id

### What I expect to happen:
Just a normal response

### What actually happened:
Sometime I get the above error, sometime it works

### What version of Python and Zope/Addons I am using:

Python3.11.10 Zope5.10

**Comment by d-maurer** [comment]
GiaBen wrote at 2024-10-30 07:28 -0700:
>## BUG/PROBLEM REPORT / FEATURE REQUEST
>File "/home/myusername/apps/zope5plus/zope5100/eggs/transaction-4.0-py3.11.egg/transaction/_transaction.py", line 570, in abort
>    raise v.with_traceback(tb)
>  File "/home/myusername/apps/zope5plus/zope5100/eggs/transaction-4.0-py3.11.egg/transaction/_transaction.py", line 551, in abort
>    rm.abort(self)
>  File "/home/myusername/apps/zope5plus/zope5100/eggs/Products.ZSQLMethods-4.1-py3.11.egg/Shared/DC/ZRDB/TM.py", line 90, in abort
>    self._abort()
>  File "/home/myusername/apps/zope5plus/zope5100/eggs/Products.ZMySQLDA-5.0-py3.11.egg/Products/ZMySQLDA/db.py", line 679, in _abort
>    self._query('ROLLBACK')
>  File "/home/myusername/apps/zope5plus/zope5100/eggs/Products.ZMySQLDA-5.0-py3.11.egg/Products/ZMySQLDA/db.py", line 519, in _query
>    self.db.query(query)
>  File "/home/myusername/apps/zope5plus/zope5100/lib/python3.11/site-packages/MySQLdb/connections.py", line 261, in query
>    _mysql.connection.query(self, query)
>MySQLdb.OperationalError: (2006, 'Server has gone away')

I doubt that this is a `ZMySQLDA` problem.
It looks more like an operational problem (your database process died
(for whatever reason)).

**Comment by GiaBen** [comment]
I guess Dieter was right. Thanks anyway!

**Comment by michaelbrunnbauer** [comment]
I'd like to reopen this because I think this is a problem with the connection pool not recovering from a lost connection. I have seen it many times now. As described above, "Sometime I get the above error, sometime it works". There never was a problem with the database and restarting Zope always made the problem go away - I suppose until MySQL closes a connection due to inactivity again (the default timeout is 28800 seconds).

**Comment by d-maurer** [comment]
Michael Brunnbauer wrote at 2024-12-14 12:52 -0800:
>I'd like to reopen this because I think this is a problem with the connection pool not recovering from a lost connection. I have seen it many times now. As described above, "Sometime I get the above error, sometime it works". There never was a problem with the database and restarting Zope always made the problem go away - I suppose until MySQL closes a connection due to inactivity again (the default timeout is 28800 seconds).

I have used `mysql` only for a short time (and then switched to `Postgres`).
But I expect that `mysql` behaves similar to `Postgres`: i.e.
it creates a new process for each incoming connection.
When this (connection) process should die, then the connection
likely reports something like "server has gone away").
It is not necessary that the main database process is in any trouble;
then new connections will be established normally.

I suggest you look into the database logfiles for any hints
towards unexpected process terminations.

**Comment by michaelbrunnbauer** [comment]
As I said, a new connection is not established normally by ZMySQLDA / Zope. Instead, I get the "server has gone away" message again and again in event.log while Zope serves an internal server error message without reporting the exception in error_log while logging the request with a 200 status code in Z4.log. All of this only if I hit the apparently broken connection in the database pool - so some requests work normally and some don't. I can assure you I've had this on multiple servers that have no issue with the database.

**Comment by dataflake** [comment]
The problem appears to be mysqlclient. I saw the same issues on a site that moved from version 1.4.6 to the current 2.2.6 release. Going back to 1.4.6 solved it.

**Comment by michaelbrunnbauer** [comment]
I am currently using 2.2.4 and first noticed this problem on Oct 22 - probably after upgrading to Zope 5.11. Maybe earlier.

**Comment by GiaBen** [comment]
It looks like limiting the threads  with the line threads = 2 in etc/zope.ini just after [server:main] setups is fixing the issue. My web hosting guy came up with this solution.

**Comment by dataflake** [comment]
@michaelbrunnbauer in the process logs I noticed that the issue seems to start with the "server went away" error coming out of the `ping` call on the connection at transaction start. Looking at the change history the C code in mysqlclient's ping method changed for the 2.2.1 release. Could you try version 2.2.0?

@GiaBen I appreciate the input, but that's a workaround and not a fix.

**Comment by michaelbrunnbauer** [comment]
2.2.0 does not have the issue. I tested by restarting the database and 2.2.4 does not recover from that while 2.2.0 does.

**Comment by dataflake** [comment]
Thanks for trying that out - I had not been able to reproduce it on my test systems and did not want to test on the production systems where I noticed it yesterday.

I would guess that the early failure in the transaction essentially wrecks it in a non-recoverable way that is so bad that other actions involving transactions - like storing error information in the error log - cannot proceed and the server comes to a standstill.

**Comment by michaelbrunnbauer** [comment]
I can't nail the culprit, though, as I can only test with my full Zope instance. It might be mysqlclient or some component further up mishandling the change introduced in mysqlclient.

**Comment by dataflake** [comment]
I'm thinking the changed mysqlclient connection `ping` method means some implicit assumptions here are no longer fulfilled. The Zope transaction-related code clearly never expected an exception to be thrown inside `ping`. The trouble always starts like this inside `Products.ZMySQLDA.db.DB._begin`, which ties the Zope and MySQL/MariaDB transactions, one of the worst places for that to happen:

```
2024-12-13 14:18:48,683 ERROR   [ZMySQLDA:650][waitress-1] exception during _begin
Traceback (most recent call last):
  File "/opt/zope/.eggs/Products.ZMySQLDA-5.0-py3.11.egg/Products/ZMySQLDA/db.py", line 644, in _begin
    self.db.ping()
MySQLdb.OperationalError: (2006, 'Server has gone away')
```

**Comment by d-maurer** [comment]
Michael Brunnbauer wrote at 2024-12-15 06:29 -0800:
>I can't nail the culprit, though, as I can only test with my full Zope instance. It might be mysqlclient or some component further up mishandling the change introduced in mysqlclient.

Apparently, something uses the `poll` to check whether the connection
(e.g. from a pool) is still functional and this seems no longer
to work as before.

I would investigate what `poll` now returns in case of a broken
connection and adapt accordingly.

**Comment by d-maurer** [comment]
Jens Vagelpohl wrote at 2024-12-15 07:02 -0800:
>I'm thinking the changed mysqlclient connection `ping` method means some implicit assumptions here are no longer fulfilled. The Zope transaction-related code clearly never expected an exception to be thrown inside `ping`. The trouble always starts like this inside `Products.ZMySQLDA.db.DB._begin`, which ties the Zope and MySQL/MariaDB transactions, one of the worst places for that to happen:

There is code to handle a "hosed connection" at many places.
Maybe, we should interpret a failing "ping" as an indication for
a "hosed connection" and force a reconnection in this case.

**Comment by dataflake** [comment]
@michaelbrunnbauer Could you test against the branch from https://github.com/zopefoundation/Products.ZMySQLDA/pull/36 in conjunction with `mysqlclient` version 2.2.6?

**Comment by d-maurer** [comment]
Jens Vagelpohl wrote at 2024-12-16 05:54 -0800:
>@michaelbrunnbauer Could you test against the branch from https://github.com/zopefoundation/Products.ZMySQLDA/pull/36 in conjunction with `mysqlclient` version 2.2.6?

When I have understood this right, Michael has
provoked the problem by shutting down the database.
In this case, the reconnection will fail.
A test would need to restart the database (such that old
connection get "hosed" but a reconnect is possible).

**Comment by michaelbrunnbauer** [comment]
I applied the changes to db.py and the problem is gone with myqlcient 2.2.4 and 2.2.6.

**Comment by dataflake** [comment]
@d-maurer Michael said he restarted the database, he did not just shut it down.

Thank you, Michael!

**Comment by dataflake** [comment]
I have just released version 6.0 with a fix that tries to prevent unrecoverable errors due to stale connections at the start of a Zope transaction.
